File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

  sumOfMultiples :: Integral a => [a] -> a -> a
  sumOfMultiples factors limit =
--    sum $ filter isMultiple [1..pred limit]
--    sum multiples
    sum `of'` multiples
    where
      multiples = [ i | i <- [1..pred limit],  isMultiple i]
      isMultiple x = any (isDivisibleBy x) factors
      isDivisibleBy x = (== 0) . (mod x)
      of' = ($)  -- silly humor

File: java/sum-of-multiples/src/main/java/SumOfMultiples.java
import java.util.Arrays;
import java.util.function.IntPredicate;
import java.util.function.Supplier;
import java.util.stream.IntStream;

public class SumOfMultiples {

    public static int sum(int limit, int[] factors) {
        Supplier<IntStream> factorStream =
            () -> Arrays.stream(factors);

        IntPredicate isMultiple =
            x -> factorStream.get()
                   .anyMatch(factor -> x % factor == 0);

        return IntStream.range(1, limit)
                 .filter(isMultiple)
                 .sum();
    }
}
